src/core/components/pin/gc-lock.js

```diff
@@ -54,26 +48,30 @@ class PullLocker {
   }
 
   take () {
-    return pull(
-      pullAsyncMap((i, cb) => {
-        if (this.lockRequested) {
-          return cb(null, i)
-        }
-        this.lockRequested = true
-
-        this.emitter.emit(`${this.type} request`)
-
-        this.mutex[this.type]((releaseLock) => {
-          cb(null, i)
-          this.releaseLock = releaseLock
-        })
+    return pullAsyncMap((i, cb) => {
+      // Check if the lock has already been acquired.
+      // Note: new items will only come through the pull stream once the first
+      // item has acquired a lock.
+      if (this.releaseLock) {
+        // The lock has been acquired so return immediately
+        return cb(null, i)
+      }
+
+      // Request the lock
+      this.mutex[this.type]((releaseLock) => {
+        // The lock has been granted, so run the locked piece of code
+        cb(null, i)
+
+        // Save the release function to be called when the stream completes
+        this.releaseLock = releaseLock
       })
-    )
+    })
   }
 
   // Releases the lock
   release () {
     return pullThrough(null, (err) => {
+      // When the stream completes, release the lock
       this.releaseLock(err)
```

Comment:
this doesn't need to be protected ? like

```
if(this.releaseLock) {
	this.releaseLock(err)
	this.releaseLock = null
}else{
	throw new Error('No lock to release')
}
```

Reply:
Because `release()` is called after the stream completes, it shouldn't be possible to get here without a lock having been acquired (and `this.releaseLock` being set)

Reply:
what happen if `lockedPullFn()` doesnt work properly, throws or break flow ?

Reply:
`lockedPullFn()` returns a pull stream, so it depends on what that pull-stream does

Reply:
what im asking is if `lockedPullFn()` does something wrong is this code still safe ?

Reply:
I don't believe that will affect anything in the `PullLocker.release()` function - the second callback to pullThrough() should only be invoked once the stream ends.

If an exception is thrown from within the pull-stream returned by lockPullFn() I believe it will be thrown outside the control of the locking class.